parquet/src/main/java/org/apache/iceberg/parquet/PruneColumns.java

```diff
@@ -45,13 +44,16 @@ public Type message(MessageType message, List<Type> fields) {
     for (int i = 0; i < fields.size(); i += 1) {
       Type originalField = message.getType(i);
       Type field = fields.get(i);
-      if (selectedIds.contains(getId(originalField))) {
-        builder.addField(originalField);
-        fieldCount += 1;
-      } else if (field != null) {
-        builder.addField(field);
-        fieldCount += 1;
-        hasChange = true;
+      Integer fieldId = getId(originalField);
+      if (fieldId != null) {
+        if (selectedIds.contains(fieldId)) {
```

Comment:
Why is this not `if (fieldId != null && selectedIds.contains(fieldId))`?

The else case is used when a sub-field is projected by ID. So the question is whether a sub-field can be projected if its parents aren't mapped. I think we should allow it because it would be confusing to have a value mapped, but still get nulls because a parent is not.

Reply:
Make sense to me. Updated.